From fe8eb5dacf5b3c805ca7f0974aeab019e5d1dffc Mon Sep 17 00:00:00 2001
From: Rob Herring <r.herring@freescale.com>
Date: Tue, 21 Apr 2009 09:39:44 -0500
Subject: [PATCH] ENGR00111670-2 mxcfb: support mode string on command line

Support mode string on kernel cmd line.
Unblank DI1 when it is primary display.

Signed-off-by: Rob Herring <r.herring@freescale.com>
---
 arch/arm/plat-mxc/include/mach/mxc.h |    1 +
 drivers/video/mxc/mxc_ipuv3_fb.c     |   10 +++++++++-
 2 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/arch/arm/plat-mxc/include/mach/mxc.h b/arch/arm/plat-mxc/include/mach/mxc.h
index a40b2d8..2668628 100644
--- a/arch/arm/plat-mxc/include/mach/mxc.h
+++ b/arch/arm/plat-mxc/include/mach/mxc.h
@@ -151,6 +151,7 @@ struct mxc_lightsensor_platform_data {
 
 struct mxc_fb_platform_data {
 	struct fb_videomode *mode;
+	char *mode_str;
 	u32 interface_pix_fmt;
 };
 
diff --git a/drivers/video/mxc/mxc_ipuv3_fb.c b/drivers/video/mxc/mxc_ipuv3_fb.c
index 55911d8..2b9d43d 100644
--- a/drivers/video/mxc/mxc_ipuv3_fb.c
+++ b/drivers/video/mxc/mxc_ipuv3_fb.c
@@ -901,13 +901,14 @@ static int mxcfb_probe(struct platform_device *pdev)
 		if (!g_dp_in_use) {
 			mxcfbi->ipu_ch_irq = IPU_IRQ_BG_SYNC_EOF;
 			mxcfbi->ipu_ch = MEM_BG_SYNC;
+			mxcfbi->blank = FB_BLANK_UNBLANK;
 		} else {
 			mxcfbi->ipu_ch_irq = IPU_IRQ_DC_SYNC_EOF;
 			mxcfbi->ipu_ch = MEM_DC_SYNC;
 			fbi->var.nonstd = IPU_PIX_FMT_UYVY;
+			mxcfbi->blank = FB_BLANK_POWERDOWN;
 		}
 		mxcfbi->ipu_di = pdev->id;
-		mxcfbi->blank = FB_BLANK_POWERDOWN;
 
 		strcpy(fbi->fix.id, "DISP3 BG - DI1");
 	} else if (pdev->id == 2) {	/* Overlay */
@@ -942,6 +943,13 @@ static int mxcfb_probe(struct platform_device *pdev)
 	fbi->var.xres = 240;
 	fbi->var.yres = 320;
 
+	if (!fb_mode && plat_data && plat_data->mode_str)
+		fb_mode = plat_data->mode_str;
+
+	if (fb_mode)
+		fb_find_mode(&fbi->var, fbi, fb_mode, NULL, 0, NULL,
+			     default_bpp);
+
 	if (plat_data) {
 		mxcfbi->ipu_di_pix_fmt = plat_data->interface_pix_fmt;
 		if (plat_data->mode)
-- 
1.5.4.4

